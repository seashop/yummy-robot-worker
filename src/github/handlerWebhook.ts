import type { LarkRobot } from "../lark/robot";
import type {
  WebhookEventMap,
  WebhookEventName,
  PushEvent,
  WorkflowJobEvent,
} from "@octokit/webhooks-types";

/**
 * See https://open.feishu.cn/tool/cardbuilder?from=howtoguide
 */
export const makeInteractiveCard = ({
  title,
  content,
  url,
  at,
  atAll,
}: {
  title: string;
  content: string;
  url?: string;
  /**
   * Custom bot only supports `@users` using open_id;
   */
  at?: string[];
  atAll?: boolean;
}) =>
  ({
    config: {
      wide_screen_mode: true,
    },

    header: {
      template: "purple",
      title: {
        tag: "plain_text",
        content: title,
      },
    },

    elements: [
      {
        tag: "markdown",
        content,
      },

      at && {
        tag: "div",
        text: {
          content: at.map((email) => `<at email=${email}></at>`).join(" "),
          tag: "lark_md",
        },
      },

      atAll && {
        tag: "div",
        text: {
          content: "<at id=all></at>",
          tag: "lark_md",
        },
      },

      url && {
        actions: [
          {
            tag: "button",
            text: {
              content: "立即查看",
              tag: "plain_text",
            },
            type: "primary",
            url,
          },
        ],
        tag: "action",
      },
    ].filter(Boolean),
  } as const);

export const handleWebhook = async (
  name: WebhookEventName,
  event: WebhookEventMap[WebhookEventName],
  robot: LarkRobot
) => {
  let card = null;
  let e = null;
  switch (name) {
    case "push":
      e = event as PushEvent;
      card = makeInteractiveCard({
        title: `${e.repository.name} 有新的推送`,
        content: [
          `repository: ${e.repository.full_name}`,
          `branch : ${e.ref}`,
          `committer: ${e.pusher.name}`,
          `compare: ${e.compare}`,
        ]
          .filter(Boolean)
          .join("\n"),
      });
      break;
    case "workflow_job":
      e = event as WorkflowJobEvent;
      card = makeInteractiveCard({
        title: `${e.repository.name} 有新的Job`,
        content: [`repository: ${e.repository.full_name}`]
          .filter(Boolean)
          .join("\n"),
      });
      break;
      break;
    default:
      console.error("Unknown event:", event);
      break;
  }

  if (card) {
    const resp = await robot.send({
      msg_type: "interactive",
      card,
    });
    return resp;
  }

  console.error("Unknown event:", event);
};
