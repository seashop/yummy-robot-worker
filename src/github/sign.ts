import hexToArrayBuffer from "hex-to-array-buffer";

async function verifybWebhookSignature(
  secret: string,
  payload: BufferSource,
  sign: string
) {
  if (!payload) return null;

  const enc = new TextEncoder();
  const algorithm = { name: "HMAC", hash: "SHA-256" };

  const key = await crypto.subtle.importKey(
    "raw",
    enc.encode(secret),
    algorithm,
    false,
    ["sign", "verify"]
  );

  // const sign = await crypto.subtle.sign(algorithm.name, key, payload)

  return await crypto.subtle.verify(
    algorithm.name,
    key,
    hexToArrayBuffer(sign.replace(/^sha256=/, "") ?? ""),
    payload
  );
}

// Verifies authenticity of a webhook event.
// https://developer.github.com/webhooks/#delivery-headers
export async function verifyWebhookEvent(
  secret: string,
  headers: Headers,
  payload: BufferSource
) {
  if (typeof secret !== "string") {
    const errMsg = "Must provide a 'GITHUB_WEBHOOK_SECRET' env variable";
    return { statusCode: 401, body: errMsg };
  }

  const sig =
    headers.get("x-hub-signature-256") || headers.get("X-Hub-Signature-256");
  if (!sig) {
    const errMsg = "No X-Hub-Signature-256 found on request";
    return { statusCode: 401, body: errMsg };
  }

  const githubEvent =
    headers.get("x-github-event") || headers.get("X-GitHub-Event");
  if (!githubEvent) {
    const errMsg = "No X-Github-Event found on request";
    return { statusCode: 422, body: errMsg };
  }

  const id =
    headers.get("x-github-delivery") || headers.get("X-GitHub-Delivery");
  if (!id) {
    const errMsg = "No X-Github-Delivery found on request";
    return { statusCode: 401, body: errMsg };
  }

  const isValid = await verifybWebhookSignature(secret, payload, sig);
  if (!isValid) {
    const errMsg =
      "X-Hub-Signature-256 incorrect. Github webhook token doesn't match";
    return { statusCode: 401, body: errMsg };
  }

  return null;
}
