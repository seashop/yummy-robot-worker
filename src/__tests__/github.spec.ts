import type { WebhookEvent } from "@octokit/webhooks-types";
import { describe, expect, vi, test } from "vitest";
import { handleWebhook } from "../github/handlerWebhook";
import type { LarkRobot } from "../lark/robot";
import payload from "./fixtures/push.json";

describe("gitlab", () => {
  test("should send data match snapshot", async () => {
    const mockRobot: LarkRobot = {
      send: vi.fn((...args) => {
        expect(args).toMatchSnapshot();
        return Promise.resolve({ code: 1, msg: "" });
      }),
    };
    await handleWebhook("push", payload as WebhookEvent, mockRobot);
    expect(mockRobot.send).toBeCalledTimes(1);
  });
});
