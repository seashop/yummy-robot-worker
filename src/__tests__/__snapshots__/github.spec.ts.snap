// Vitest Snapshot v1

exports[`gitlab > should send data match snapshot 1`] = `
[
  {
    "content": {
      "text": "repository: seashop/yummy-robot-worker
branch : refs/heads/yummy
committer: xqbumu
compare: https://github.com/seashop/yummy-robot-worker/compare/91648ad0d860...e218cd90c429",
    },
    "msg_type": "text",
  },
]
`;
